authenticated message subscription endpoints

jagerman opened the pull request:
This PR adds endpoints to storage server that allows someone to maintain a oxenmq connection to one or more swarm members and receive pushed notifications through that connection when new messages are delivered.

This is authenticated: subscribing requires a signature from the mailbox owner signed within the past 14 days, and connections have to be refreshed at least once/hour to keep the push notifications alive.

The immediate use for this will be for more efficient push notifications for mobile clients using the push notification server, but this mechanism will eventually also allow clients (over lokinet) to get messages pushed to them rather than having to frequently poll.

jagerman committed:
Message monitoring

This PR adds endpoints to storage server that allows someone to maintain
a oxenmq connection to one or more swarm members and receive pushed
notifications through that connection when new messages are delivered.

This is authenticated: subscribing requires a signature from the mailbox
owner signed within the past 14 days, and connections have to be
refreshed at least once/hour to keep the push notifications alive.

The immediate use for this will be for more efficient push notifications
for mobile clients using the push notification server, but this
mechanism will eventually also allow clients (over lokinet) to get
messages pushed to them rather than having to frequently poll.

Flam3z commented:
may I ask what this will be for regarding session?

KeeJef commented:
Its an update for how we manage push notifications, basically refactoring Oxen storage server to notify the push notification server of any new messages for users who are subscribed to use push notifications. "Message monitoring" sounds scary but its really just monitoring to ensure PN's get set properly. Storage servers still cant read message contents or anything like that.

majestrate commented:
On Mon, 25 Jul 2022 19:58:57 -0700
Flamez ***@***.***> wrote:

> may I ask what this will be for regarding session?
> 

this cuts the PN latency by 60% (it is 1/3 the rtt as a result)

jagerman committed:
Bump oxenc for bt_dict_producer fix

jagerman committed:
Add testnet network tests for monitoring endpoint

Flam3z commented:
> Its an update for how we manage push notifications, basically refactoring Oxen storage server to notify the push notification server of any new messages for users who are subscribed to use push notifications. "Message monitoring" sounds scary but its really just monitoring to ensure PN's get set properly. Storage servers still cant read message contents or anything like that.

Thanks for responding and yeah I did kind of think it was what the PR meant :P

jagerman commented:
> Thanks for responding and yeah I did kind of think it was what the PR meant :P

I reworded it to be less "scary"

jagerman committed:
Allow subscribing to multiple pubkeys at once

The PN server potentially needs to subscribe to many addresses at once;
this allows it to do so in a single request rather than needing many
separate small requests.

jagerman committed:
Add test case for multi-sub request

jagerman committed:
Clarify "subkey" vs "subkey tag" descriptions

majestrate commented on oxenss/server/omq.h:
```diff
@@ -130,8 +130,8 @@ class OMQ {
     ///     signature uses *this* key, not the derived x25519 key.
-    /// - S -- (optional) a 32-byte authentication subkey to use for authentication.  The
-    ///   signature with such a subkey uses a derived key (as described in the RPC endpoint
+    /// - S -- (optional) a 32-byte authentication subkey tag to use for authentication.  The
```
this sounds a bit more ambiguous as "tags" are also used in authenticated encryption as authentication tags.

jagerman replied:
True, but the next line is "go see the rpc docs" which has a paragraph about the subkey value.

jagerman committed:
Merge namespace/data values into a single subscription

Maintaining multiple subscriptions for the same account on the same
connection for each different set of namespaces/want_data is painful, so
merge them into one.

This has some implications about allowing a single connection to renew
namespace and want_data subscriptions (even if those namespaces and/or
want_data flag aren't getting specified in renewals), but that is
relatively minor and it allows considerable simplification of
subscriptions.
